### ISSUE TITLE
<video /> attribute needed but not guaranteed by React

### ISSUE THREAD

**Issue by elrumordelaluz** [facebook/react#10389]
As @gaearon [mentioned](https://twitter.com/dan_abramov/status/893511715463340033), React does not guarantee an *attribute* will be set, so probably this is not a bug.

If I understand well, react will ensure the property is set anyway. 

**Current behaviour**
React renders the html `<video />` element without the attribute `muted` when explicitly passed.

**Demo time**
In [this pen](https://codepen.io/elrumordelaluz/pen/ba4b5294cec7ea455c66e9b9d0a6c889?editors=0010) I made a simple example setting `muted` to the element and obtaining the result below:
![pen-screen-shot](https://user-images.githubusercontent.com/784056/28986491-7b2dd45e-7967-11e7-9f63-b4db88db9989.png)

Actually the property is set well, since the [original medial file](http://clips.vorwaerts-gmbh.de/VfE_html5.mp4) has an audio track and in the pen result is muted.

**The point**
I think is most a _specific need_ than the _expected behaviour_.
From the functionality POV, it is absolutely ok, my _Component_ renders a `<video />` muted as requested and so on.

But there are _browsers_ and _policies_, more specifically related to this issue, Webkit and the [New <video> Policies for iOS](https://webkit.org/blog/6784/new-video-policies-for-ios/) updated one year ago, with some interesting changes for the `<video />` element.
The part interested is
> `<video muted>` elements will also be allowed to autoplay without a user gesture.

So, the _specific need_ is to have the _explicit_ `attribute` to tell the browser that this video could be `autoPlay`ed.

There's a similar [issue](https://github.com/facebook/react/issues/6544)

**Comment by syranide** [comment]
The `muted` **property** is live just like `value` for inputs and reflects the current value. But the `muted` attribute reflects the initial state of the component.

So I think the current behavior is wrong, because `<video muted />` should force it to always be muted, just like `<input value="foo" />` forces it to always be `"foo"`. So one way to resolve this is to start from scratch; introduce `defaultMuted` property which maps to the attribute only. Then there should be a `muted` property which is controlled (just like input `value`) and there needs to be a callback to go with that e.g. `onMutedChange`. However! Mute is just equal to volume being zero, so the DOM only exposes the `volumechange` event.

So perhaps a more DOM-truthful and minimal implementation would be to have `muted` be the attribute and reflect only the initial mute state. With this you can then manually control the video element via JS (like we recommend elsewhere with React).

Or React can take it further and also support a special-property `volume` and its corresponding event `onVolumeChange`. But this is a slippery slope and there are tons of such properties that should be added then, which also collides with attribute behaviors. Elsewhere we recommend that people simply manually set properties via JS to control elements and I think the same general recommendation works best here too, manually set the properties and expose your fancy video player as a React component if you want to reuse it.

**tl;dr** Yes, `muted` should set the attribute only, so that it only sets the initial state of the element.

**Comment by montogeek** [comment]
I am facing the same issue with Android 4.4, according to new (from a year ago), `muted` attribute needs to be present.

**Comment by syranide** [comment]
@gaearon This should be an easy fix if you're OK with the direction (making `muted` set the attribute instead), although there is potential to break some existing code if someone decided to implement their own mute button and rely on this prop.

**Comment by aweary** [comment]
Since `defaultMuted` is already a spec-defined property reflecting the `muted` attribute, it would make sense to let users utilize this API. The React API is already property-focused, so it's not a big leap. This would be the least complex change, and shouldn't be breaking.

**Comment by kiurchv** [comment]
@aweary Any progress on this?

**Comment by NicolasPoirierB** [comment]
@aweary Same question here, we need the muted attribute to be passed to the video tag to allow autoplaying on mobile devices. Has there been any progress in this direction?

Thank you!

**Comment by jquense** [comment]
folks progress on an Issue is generally visible and takes the form of a PR. If you don't see a linked PR or  updates on an issue assume there is no more progress. 

It's unclear what the path forward is for this to me. We don't really have way of saying "sync the attribute once and then use the property" and I don't know that adding said logic is even a good idea...

**Comment by syranide** [comment]
> It's unclear what the path forward is for this to me. We don't really have way of saying "sync the attribute once and then use the property" and I don't know that adding said logic is even a good idea...

@jquense I think I understand your point, but inputs have both `value` and `defaultValue`. This is exactly the same thing, there should probably be a `defaultMuted` and `muted`. It's a "live" value and needs to be special-cased in the same way, that or otherwise just have `muted` prop which sets the attribute but that means users need to manually access the `muted` property if they want to control it precisely (just like you can with `defaultValue` for inputs), which is also fine.

Personally, just having `muted` set the attribute and leaving everything else up to the user is perfectly fine (and IMHO how input should have been implemented too).

**Comment by aweary** [comment]
I'm hesitant to change `muted` to act like `value`; we want to get away from that special-cased behavior, since it's caused a number of problems. Of course, `muted` would likely be much less problematic since it's just a boolean value.

> We don't really have way of saying "sync the attribute once and then use the property" and I don't know that adding said logic is even a good idea...

`defaultMuted` is the spec'd solution for this. It sets the initial muted value by reflecting the `muted` attribute, which can then be controlled using the `muted` property.

We don't support many attributes that aren't actually DOM attributes, but we do have `defaultValue` so I think there's a good argument for mirroring that API here.

**Comment by jquense** [comment]
> we want to get away from that special-cased behavior, since it's caused a number of problems.

yeah the controlling behavior ends up causing so many headaches i wouldn't want to do it again, though i agree that it's the right behavior generally.

> We don't support many attributes that aren't actually DOM attributes, but we do have defaultValue so I think there's a good argument for mirroring that API here.

I like that idea, tho It's always a slippery slope, what would the SSR behavior if someone specifies `defaultMuted` but no `muted`, ? I suppose we already handle this with value and checked...

**Comment by aweary** [comment]
> I like that idea, tho It's always a slippery slope, what would the SSR behavior if someone specifies defaultMuted but no muted, ? I suppose we already handle this with value and checked...

Yeah, I imagine it would behavior [just like `defaultValue` does for SSR](https://codesandbox.io/s/ym0oqo11qj), by rendering the `value` attribute with the provided value.

**Comment by jquense** [comment]
ya my concern is just that most of the value/defaultValue is special cased through the codebase sort of assuming the pattern will be used just for inputs. I should probably go see how it works tho before worrying about that...

**Comment by aweary** [comment]
My assumption is that the one use case for `defaultMuted` is to handle the use case where the `muted` attribute should be present when the video is first rendered. Since it has no dynamic behavior, per the spec, I think just defining it as a boolean attribute that must set the property (easy to do with `PropertyInfoRecord`) would be good enough? I hope at least 😅

**Comment by mystrdat** [comment]
I don't give a 💩 what React thinks about the specs, I want to see an attribute I explicitly declare rendered.

**Comment by vinay72** [comment]
`const { Component } = React
const { render } = ReactDOM

class App extends Component {
  render () {
    return (
      <video style={{ width: '100%' }}
    src="http://clips.vorwaerts-gmbh.de/VfE_html5.mp4" 
        autoPlay
        video />
    )
  }
}

render(<App />, document.querySelector("#main"))`
If you run this code there is a video playing automatically with audio.

**Comment by ctate** [comment]
`muted` is an **extremely** important attribute for background videos. _Please_ expose it to the browser. 🙏

In the meantime, if you need to autoplay background videos, try this:

```
import * as React from 'react';
import { Component } from 'react';

export class VideoComponent extends Component {
  videoContainer: HTMLDivElement;
  componentDidMount() {
    const video = document.createElement('video');
    video.autoplay = true;
    video.muted = true;
    video.loop = true;

    const source = document.createElement('source');
    source.src = '/path/to/your/video.mp4';
    source.type = 'video/mp4';
    video.append(source);

    this.videoContainer.append(video);
  }
  render() {
    return (
      <div ref={(ref) => { this.videoContainer = ref; }} />
    );
  }
}
```

Thank you for your consideration.

**Comment by tomasdev** [comment]
Indeed, ended up doing dangerouslySetHTML because React doesn't accept muted.

**Comment by Zemnmez** [comment]
https://www.npmjs.com/package/react-html5video this component also seems to expose `muted`

**Comment by ml242** [comment]
And still this issue persisted

**Comment by coalescecreate** [comment]
While muted still doesn't show up as an attribute on the video element. Simply adding the playsInline makes the video play on iOS/Safari and Chrome.
```
<video className={s.video} src={video.url} autoPlay={video.autoplay} playsInline muted={video.mute ? 1 : null} loop={video.loop} />
```

**Comment by adjohu** [comment]
I feel like this is a source of bugs and therefore not really a feature request. Is there any plan to work on this? None of the workarounds feel nice to me.

**Comment by keithpickering** [comment]
Agree with @adjohu, this is a bug, not a feature request. Video elements will not autoplay on in mobile Safari without the `muted` attribute. How hard can it be to just guarantee the attribute? I can use `dangerouslySetInnerHTML` for the video element, but I also need a ref to it, so that doesn't work.

**Comment by joshverd** [comment]
Almost two years and there still is no good way of doing this :/

**Comment by keithpickering** [comment]
@joshverd is there ANY way of doing it? Haha. I can't find anything about a video element in react, with a muted attribute, and also with a ref. It's like I'm on Mars, I can't believe this is such an uncommon requirement 😬

**Comment by joshverd** [comment]
@keithpickering 
Actually, I got something working. It's really hack-ish but it works. 

Basically, you want to save the element's 'ref' 
`<video ref={ref => this.videoRef = ref} src="link/to/video.mp4" />`

Then you have to set the element's muted property to true and play the video all from within JS. 
`startVideoPlayback() {
  // Muting the video
  this.videoRef.muted = true;

  // Playing the video
  this.videoRef.play();
}`

This is the only way I can get Chrome to "autoplay" video, but you still have to call the 'startVideoPlayback' function. Anything else and it gets paused by Chrome's autoplay detection.

**Comment by keithpickering** [comment]
Wow, that `.play()` works on mobile? I can never keep track of this dumb spec lol. That's an interesting approach, thanks!

**Comment by olarsson** [comment]
Still experiencing this issue in React 16.11.0, muted attribute not showing up on the video element. Did I miss something?

**Comment by mystrdat** [comment]
You missed the entire 2 years of React team's incompetence in following the status quo.

**Comment by dimitarnestorov** [comment]
Still a valid issue

**Comment by ml242** [comment]
updated a fiddle to reflect React 16 since the OP was still using  15.

still valid.

demo: https://jsfiddle.net/matt1234/vqnrc83a/1/

**Comment by mar753** [comment]
Isn't it related with that: https://developers.google.com/web/updates/2017/09/autoplay-policy-changes

When you are using e.g. jsfiddle.net to test your code snippet, everything is rendered inside `iframe` there and the default behavior for e.g. Chrome is to block autoplay of such videos. If you would like the autoplay to work, you can add something like this to your local application:
`<iframe src="http://clips.vorwaerts-gmbh.de/VfE_html5.mp4" allow="autoplay" />`

**Comment by mar753** [comment]
If you would like to have `muted` attribute present in HTML you can use this component (hook version) as well:
```
import { useRef, useEffect } from 'react'

const Player = () => {
  const videoRef = useRef(null)

  useEffect(() => {
    const { current: videoElement } = videoRef
    videoElement.setAttribute('muted', '')
  }, [videoRef])

  return (
    <video
      src="http://clips.vorwaerts-gmbh.de/VfE_html5.mp4"
      ref={videoRef}
      autoPlay
      playsInline
      muted
    />
  )
}

export default Player
```

Regarding autoplay:
Isn't it related with that: https://developers.google.com/web/updates/2017/09/autoplay-policy-changes

When you are using e.g. jsfiddle.net to test your code snippet, everything is rendered inside `iframe` there and the default behavior for e.g. Chrome is to block autoplay of such videos. If you would like the autoplay to work, you can add something like this to your local application:
`<iframe src="http://clips.vorwaerts-gmbh.de/VfE_html5.mp4" allow="autoplay" />`

**Comment by mar753** [comment]
If you would like to have `muted` attribute present in HTML you can use this component (hook version) as well:
```
import { useRef, useEffect } from 'react'

const Player = () => {
  const videoRef = useRef(null)

  useEffect(() => {
    const { current: videoElement } = videoRef
    videoElement.setAttribute('muted', '')
  }, [])

  return (
    <video
      src="http://clips.vorwaerts-gmbh.de/VfE_html5.mp4"
      ref={videoRef}
      autoPlay
      playsInline
      muted
    />
  )
}

export default Player
```

Regarding autoplay:
Isn't it related with that: https://developers.google.com/web/updates/2017/09/autoplay-policy-changes

When you are using e.g. jsfiddle.net to test your code snippet, everything is rendered inside `iframe` there and the default behavior for e.g. Chrome is to block autoplay of such videos. If you would like the autoplay to work, you can add something like this to your local application:
`<iframe src="http://clips.vorwaerts-gmbh.de/VfE_html5.mp4" allow="autoplay" />`

**Comment by ml242** [comment]
@mar753 - that's an elegant solution using hooks but it doesn't answer the question of why this prop isn't being passed, despite being a known bug for years.

**Comment by xitanggg** [comment]
Any update on this?

**Comment by Losses** [comment]
@ gaearon sorry to borther you, but someone should handle this problem since it's really annoying.

**Comment by Losses** [comment]
@gaearon sorry to borther you, but someone should handle this problem since it's really annoying.

**Comment by boltcoder** [comment]
People who are just trying to autoplay muted <video> (usually in banner background  ). I wrote a medium article which can help you hours of googling. 
https://medium.com/@BoltAssaults/autoplay-muted-html5-video-safari-ios-10-in-react-673ae50ba1f5
Hope this helps.

**Comment by maximilientyc** [comment]
Thank you for your blog @boltcoder, but to me using `dangerouslySetInnerHtml` was not the solution I was expecting. I am expecting my `muted` attribute to be reflected inside the dom, I think browsers also expect that.

**Comment by maximilientyc** [comment]
From Google documentation about [Autoplay policy changes](https://developers.google.com/web/updates/2017/09/autoplay-policy-changes) (2017-09):

```
One cool way to engage users is about using muted autoplay and let them chose to unmute (see code snippet below). Some websites already do this effectively, including Facebook, Instagram, Twitter, and YouTube.

<video id="video" muted autoplay>
<button id="unmuteButton"></button>

<script>
  unmuteButton.addEventListener('click', function() {
    video.muted = false;
  });
</script>
```

The example speaks by itself: chrome is expecting the dom to contain the `muted` attribute, just like it's expecting it to contain the `autoplay` attribute.

**Comment by boltcoder** [comment]
Ofcourse it's not the right solution. But that's what I had to do to make it work in React. I'm just helping people out on tight deadlines to circumvent the limitations while react fixes (or doesn't fix the attribute to dom guarantee issue).

**Comment by lucashfreitas** [comment]
After 3 years, this issue is still unresolved :(

I believe this is quite important for who is using background videos.

**Comment by nelsonvassalo** [comment]
Still a bug, can't believe I was just forced to use `dangerouslySetHTML` to get around this…

**Comment by cranberrymuffin** [comment]
> Still a bug, can't believe I was just forced to use `dangerouslySetHTML` to get around this…

you dont need to use dangerouslySetHTML, try this: https://github.com/facebook/react/issues/6544#issuecomment-705702546

**Comment by gaearon** [comment]
What is the latest proposal in this thread? I think the issue has stalled because there was no single proposal that people could get behind. It would help to flesh out what exactly the behavior should be with regards to whether it should be setting an attribute or a property, and whether it should be controlled or uncontrolled. Once we have a proposal, it's easier for someone to make a PR.

**Comment by ml242** [comment]
proposal: the muted attribute when passed as a prop to a video component with a boolean value will be added to the dom.

**Comment by gaearon** [comment]
What is your plan for `defaultMuted` vs `muted`, controlled vs uncontrolled, server rendering?

**Comment by ml242** [comment]
incremental progress :)

**Comment by gaearon** [comment]
Sorry, I appreciate your intent but this is not helpful and does not move the thread forward. This thread has stalled because it doesn't have a concrete proposal. If you don't have a concrete technical proposal (what you posted is vague), let's wait for somebody to post theirs.

**Comment by ml242** [comment]
From the React16 intro:

In the past, React used to ignore unknown DOM attributes. If you wrote JSX with an attribute that React doesn’t recognize, React would just skip it. For example, this:

// Your code:
<div mycustomattribute="something" />
would render an empty div to the DOM with React 15:

// React 15 output:
<div />

In React 16, we are making a change. Now, any unknown attributes will end up in the DOM:

// React 16 output:
<div mycustomattribute="something" />

So does React omit 'muted' because it is known? I can't see why React needs to take into account every scenario for SSR and how to handle defaultMuted when it has an existing convention that developers should be able to follow.

**Comment by gaearon** [comment]
React, in general, needs to take every scenario into account. :-) It doesn't do that for arbitrary attributes that don't have special behavior because... they don't have special behavior. In this case, there are two different backing properties, and we need make sure that the behavior makes sense.

**Comment by gaearon** [comment]
In React, generally saying, JSX attributes specify the *current* value rather than the *initial* one.

```js
<input value={myState} />
```

means that if `myState` gets set by some other part of the code, the input value will reflect it.

So if `muted={true}` meant "*initially* muted" it would not be consistent with how other similar JSX attributes work.

Instead, to specify the initial behavior, we use properties starting with `default`: `defaultValue`, `defaultChecked`. This is because DOM properties that mirror the attribute are called with this convention.

So it seems like what we really want here is support for `defaultMuted`?

**Comment by ml242** [comment]
Works for me and helps devs to conform to mdn's recomendations [here](https://developer.mozilla.org/en-US/docs/Web/Media/Autoplay_guide).

**Comment by antoniopacheco** [comment]
I created a PR that fixes this issue: 
https://github.com/facebook/react/pull/20087

**Comment by atnpcg** [comment]
note, previous PR won't add defaultMuted, instead it put into the DOM the muted property and fixes the unit testing failing with <video muted>.
I did not implement **defaultMuted** because does not work for autoPlay property
you can see the difference:
https://jsfiddle.net/btph8mx3/

**Comment by stilllife00** [comment]
@gaearon My understanding here is that the agreed solution for this is to introduce `defaultMuted` to be consistent.
Do you think this could be a "good first issue" to tackle by looking at other `default*` props or there is more than that?
Is somebody already working in that direction?

**Comment by shavidze** [comment]
This issue is still under consideration?

**Comment by st-vinguyen** [comment]
I'm working on audio/video autoplay and I face the problem with `muted` attr.
https://developer.mozilla.org/en-US/docs/Web/Media/Autoplay_guide#autoplay_availability
It would be easy to pass the policy when doing autoplay with muted=true.

**Comment by infinite0007** [comment]
5 years now. Hopefully this will be fixed soon.

**Comment by ljharb** [comment]
bump

**Comment by jenniferabowd** [comment]
Bump

**Comment by st-vinguyen** [comment]
bump

**Comment by ljharb** [comment]
Bumps are only helpful when the stalebot comments. Other than that it’s just noise.

**Comment by BeezBeez** [comment]
Let's hope for a fix for 2030.

**Comment by gaearon** [comment]
Would someone like to submit a PR for the proposal in https://github.com/facebook/react/issues/10389#issuecomment-762197249? I'm not saying the team will necessarily accept it but it would help to get this rolling. I agree we need some fix for this.

**Comment by MustafaEminn** [comment]
Also, the audio element has the same issue.

**Comment by MatanYemini** [comment]
Bump

**Comment by derwaldgeist** [comment]
This is really ridiculous. So many years, and still no solution for mobile backdrop video?

**Comment by Himanxu1** [comment]
I want to work on this issue . Can anyone explain me in detail what is the issue?

**Comment by ardijanr** [comment]
This works for me on chrome.

```
onCanPlay={ (e) => { e.target.muted=true; e.target.play() } }

```

**Comment by solidiquis** [comment]
I had to do some god awful manual DOM manipulation to work around this issue that has been outstanding now for 5 years. Nearly gave me an aneurysm...

**Comment by eps1lon** [comment]
It seems to me that this issue is concerned with two problems:

1. fear of `<video autoplay muted />` not auto-playing because the `muted` attribute isn't set
1. add support of `defaultMuted` React prop

The implementation of `defaultMuted` wouldn't necessarily fix 1. as far as I can tell. There would still be no `muted` attribute if I want to control "mutedness". Unless the proposal is to always use `<video defaultMuted muted />` which is confusing considering how other `default*` props behave.

However, I don't think there is an issue with `<video autoplay muted />` in modern browsers using React 18. I created a codesandbox that tests different scenarios: https://codesandbox.io/s/video-autoplay-muted-7b65ib. The following browsers autoplay `<video autoPlay muted />`:

- Chrome Version 103.0.5060.53 (Official Build) (64-bit) (Linux)
- Chrome Version 103.0.5060.66 (Official Build) (64-bit) (Windows)
- Chrome Version 103.0.5060.70 (Android 12; Pixel 4a)
- Edge Version 95.0.1020.40 (Official build) (64-bit) (Windows)
- Firefox 97.0.1 (64-Bit) (64-bit) (Windows)
- Firefox 101.0.1 (64-bit) (Linux)

The following browsers do not autoplay this video (even if the `muted` attribute is present):

- Chrome (iOS 15.5)
- Safari (iOS 15.5)

So it would help to focus this issue if somebody could explain what concrete issue (browser + version + os) there is with not setting the `muted` attribute if we render `<video muted />`.

If there is none then we can re-focus this issue to controlled vs uncontrolled `muted` on `<video />`.

**Comment by ljharb** [comment]
@eps1lon no browser is evergreen, despite its marketing claims. Most companies' analytics still have a swath of old Chrome usage.

**Comment by aghArdeshir** [comment]
@eps1lon you can't for sure consider it as "not worth it". My scenario is I have a website builder (take Ucraft, WordPress, Wix, Webflow, etc...) in which you, as a developer, don't fully control the generated dom (structure).

Website designers can drop a video widget/element on the page, and turn on/off some controls/tweaks, like AutoPlay, Mute, Loop, Controls, etc...

The website designer turns on AutoPlay (which by the way must turn on the Mute too).

To ensure your customer's website works well on all of their customers' devices, we need to have this dom property explicitly there. As I personally faced this issue in real life, in which on an iOS device autoPlay was not working.

My workaround was:
```ts
  useEffect(() => {
    // autoplay does not work on some browsers if video does not have `muted`
    // attribute:
    // https://developer.mozilla.org/en-US/docs/Web/HTML/Element/video#attr-autoplay

    // and reacts ignores muted attribute:
    // https://github.com/facebook/react/issues/10389

    if (videoRef.current) {
      if (mute) {
        videoRef.current.setAttribute('muted', '');
      } else {
        videoRef.current.removeAttribute('muted');
      }
    }
  }, [mute]);

```

**Comment by eps1lon** [comment]
> you can't for sure consider it as "not worth it".

Sure because that's a subjective statement. But just like we don't work around browser quirks for legacy browsers like IE 11 we might not want to work around browser quirks for unmaintained Chrome versions. Especially if the issue is fixed in later versions. 

Otherwise everybody has to pay for this workaround even though they don't need to. I think it's completely fine to use your workaround if you need to support older browser versions. Arguing that usage implies the need for a fix is simply not practical considering that there are far fewer browser versions with bug fixes than browser versions without these bug fixes. So this has to be done on a case-by-case basis.

The only data point we have so far is https://github.com/facebook/react/issues/10389#issuecomment-1173052294. So if you want to contribute to this issue you might want to start collecting a browser matrix where React needs to explicitly set the `muted` attribute and where it doesn't. Then we can make an informed decision whether to fix the quirk in React.

**Comment by johiny** [comment]
I got to this issue today sure is old, I have been reading the whole day things from 2017 until now

**Comment by eps1lon** [comment]
> I got to this issue today sure is old, I have been reading the whole day things from 2016 until now has anyone got autoplay work on mobile? I have injected the muted attribute to the video tag like previous answers but it simply didn't play on mobile

@johiny  Assuming you tested this on iOS, it's a Safari issue not a React issue. This would match my testing as well:

> The following browsers do not autoplay this video (even if the `muted` attribute is present):
> 
> * Chrome (iOS 15.5)
> * Safari (iOS 15.5)

**Comment by johiny** [comment]
@eps1lon I tested it on Chrome 105 and Brave 1.43 both on (Android 10) maybe it has something to do with the manufacturer layer in this case MIUI?

**Comment by eps1lon** [comment]
> @eps1lon I tested it on Chrome 105 and Brave 1.43 both on (Android 10) maybe it has something to do with the manufacturer layer in this case MIUI?

Might be. When I tested it for Android 12 on my Pixel 4a it was working. But that was with Chrome 103.

Can you check https://7b65ib.csb.app/ and tell me which video is auto-playing?

**Comment by johiny** [comment]
@eps1lon every single one worked both on Chrome and Brave, I will have to keep testing

**Comment by eps1lon** [comment]
> @eps1lon every single one worked both on Chrome and Brave, I will have to keep testing

Then there's something off with your specific setup. This demo highlights that this issue is no longer relevant to current React versions.

**Comment by johiny** [comment]
> > @eps1lon every single one worked both on Chrome and Brave, I will have to keep testing
> 
> Then there's something off with your specific setup. This demo highlights that this issue is no longer relevant to current React versions.

yeah, I keep testing and find that something was happening with my video when it tried to start I was getting a "PIPELINE_ERROR_DECODE: video decoder reinitialization failed" error that looks like was a conflict between my video and android codecs, so if anyone is facing similar I recommend put an onError on the video can be really useful to see if something weird is happening and re-encode your video.
but thx for the help @eps1lon!

**Comment by Deeppjp116** [comment]
This can be helpful 

function VideoComponent({ muted }) {
  const videoProps = { ...(muted && { muted: true }) };

  return <video {...videoProps} />;
}

**Comment by Deeppjp116** [comment]
The `console.log()` function is used to print output to the console.

**Comment by marcwieland95** [comment]
Is there a proper solution to this problem? Also, I don't understand what the issue is with stripping out this attribute at all.

Is somebody more up-to-date on the topic to tell me what the actual issue is? A ticket since 2016, which many people seem to have a problem with, but got never addressed.

**Comment by keithpickering** [comment]
The solution is to take control of the video element in the dom since React is refusing to handle it properly. Assign a ref and call play() or pause() as needed

**Comment by idoliki** [comment]
I've just made a ref 
const isMuted = useRef(false);

then set the video component as:
<video muted={isMuted.current} ... /> 

and it shows correctly into the dom and autoplay works on iOS/macOS.

**Comment by marcwieland95** [comment]
I came up with another possible issue. The `muted` attribute sometimes doesn't get rendered. So I set it like this:

```tsx
const videoRef = useRef(null);
useEffect(() => {
    videoRef.current.defaultMuted = true;
})
```

**Comment by RsVijayTiwari** [comment]
We are come with this solution and it is working for us. But it requires dangerouslySetInnerHTML.
```jsx
<div dangerouslySetInnerHTML={{
      __html: function () {
          let videoContainer = document.createElement("video");
          let source = document.createElement("source");
          videoContainer.setAttribute("autoplay", "");
          videoContainer.setAttribute("muted", "");
          videoContainer.setAttribute("loop", "");
          source.src = "VIDEO_URL";
          videoContainer.appendChild(source);
          return videoContainer.outerHTML;
      }()
}}/>
```

**Comment by BrendanBerkley** [comment]
@eps1lon's example doesn't work in iOS because of the lack of `playsinline`, https://33v428.csb.app/ is an updated example that has it.

The updated example seems to work fine without any trickery.

**Comment by Grawl** [comment]
lol

**Comment by GrahamQuan** [comment]
for typescript / tsx 

```tsx
'use client';
import { ComponentProps, useEffect, useRef } from 'react';

export default function VideoComponent({ muted = true, ...props }: ComponentProps<'video'>) {
  const videoRef = useRef<HTMLVideoElement>(null);

  useEffect(() => {
    const videoEl = videoRef.current;

    if (!videoEl) return;
    videoEl.muted = muted; // this code
    videoEl.defaultMuted = muted; // this code

    return () => {};
  }, []);

  return <video ref={videoRef} {...props} muted={muted} />;
}

```

**Comment by codebravotech** [comment]
@GrahamQuan's solution is working for me perfectly on
Iphone: safari (latest iOS version, the phone is new), Arc (chromium), Brave
Android: Chrome, Brave 

I call it like this: 

```
          <VideoComponent
            id={videoId}
            className={cx(
              assetClasses,
              assetIsPortrait ? "max-h-[95vh]" : "max-h-[100vh]",
              "relative cursor-pointer object-scale-down",
            )}
            src={video?.asset?.url}
            autoPlay
            muted
            onContextMenu={() => false}
            loop
            playsInline
            webkit-playsinline={"true"}
          />
```

Thanks!!

**Comment by lky141319** [comment]
Claiming bounty. 
Wallet: 0xd2268DEDca16b08aF4bb6Ad20cBa877D10cf8Ac6
GitHub: @lky141319
